### PR TITLE
Tamara's Second Solution: Convert String Characters from Lowercase to Uppercase

### PR DESCRIPTION
**Pull request by tamarasaqer**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_convert_to_uppercase.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    Test cases for the `convert_to_uppercase` function.
+    """
+
+    def test_valid_strings(self):
+        """
+        Test that valid strings are correctly converted to uppercase.
+        """
+        self.assertEqual(convert_to_uppercase("hello"), "HELLO")
+        self.assertEqual(convert_to_uppercase("Tamara"), "TAMARA")
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_convert_to_uppercase.py:28]: Each unit test should have only one assertion

File: solutions/tests/test_convert_to_uppercase.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        self.assertEqual(convert_to_uppercase("123"), "123")  # Numbers remain unchanged
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_convert_to_uppercase.py:29]: This is a defensive test, already there below

File: solutions/tests/test_convert_to_uppercase.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def test_non_string_input(self):
+        """
+        Test that a non-string input raises a ValueError.
+        """
+        with self.assertRaises(ValueError):
+            convert_to_uppercase(123)
+
+        with self.assertRaises(ValueError):
+            convert_to_uppercase(None)
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_convert_to_uppercase.py:51]: Each unit test (even defensive tests) should have one assertion.

**Review by AhmedKhalifa7 — CHANGES_REQUESTED** [review]
Please, Kindly,  address these changes. if you need any help just let us know.

File: solutions/convert_to_uppercase.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    Examples:
+        >>> convert_to_uppercase("hello world")
+        'HELLO WORLD'
+        >>> convert_to_uppercase("Python")
+        'PYTHON'
+        >>> convert_to_uppercase("")
+        ''
+    """
+    if not isinstance(input_string, str):
```
Comment (AhmedKhalifa7) [inline solutions/convert_to_uppercase.py:30]: Each assumption about an argument is checked with an assertion.

**Review by AhmedKhalifa7 — CHANGES_REQUESTED** [review]

File: solutions/convert_to_uppercase.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        raise ValueError("Input must be a string.")
+    return input_string.upper()
```
Comment (AhmedKhalifa7) [inline solutions/convert_to_uppercase.py:32]: You can add a space here to show different stages of strategy

**Comment by AhmedKhalifa7** [comment]
Don't forget to label the PR. (You can use challenge or solution or both just whatever you want)

**Comment by tamarasaqer** [comment]
> Please, Kindly, address these changes. if you need any help just let us know.

File: solutions/tests/test_convert_to_uppercase.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+Date: 2025-01-10
+"""
+
+import unittest
+import sys
+import os
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+
+from convert_to_uppercase import convert_to_uppercase
```
Comment (dadishimwe) [inline solutions/tests/test_convert_to_uppercase.py:14]: from ..convert_to_uppercase

File: solutions/tests/test_convert_to_uppercase.py (hunk 3 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_convert_to_uppercase.py:64]: Everything looks good now

File: solutions/tests/test_convert_to_uppercase.py (hunk 1 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_convert_to_uppercase.py:32]: Everything looks good

File: solutions/convert_to_uppercase.py (hunk 5 above)
Comment (AhmedKhalifa7) [reply solutions/convert_to_uppercase.py:35]: Everything looks good, also I liked adding some comments

File: solutions/tests/test_convert_to_uppercase.py (hunk 2 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_convert_to_uppercase.py]: Everything looks good now

File: solutions/tests/test_convert_to_uppercase.py (hunk 6 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_convert_to_uppercase.py:14]: > from ..convert_to_uppercase

No need, she used another method

**Review by AhmedKhalifa7 — APPROVED** [review]
semplicemente perfetto👏, great job Tamara.
:shipit: !!!
